### PR TITLE
Add environment vars expansion and pre defined git hashes in the yaml

### LINKED ISSUE #116: Placeholders break soopervisor
In Ploomber, we have placeholders that users can put in their `pipeline.yaml`. For example, to index experiments by `{{git_hash}}` or timestamp with `{{now}}`. However, the value of such placeholders is computed when loading the `pipeline.yaml`.

This works fine in Ploomber since the pipelines execute in a single machine; however, when using soopervisor, this breaks. For example, the `{{git_hash}}` isn't available because we don't copy the git repository, only the existing files. And `{{now}}` breaks because it's evaluated for each task. Hence, each one sees a different timestamp.

We can solve this by resolving placeholders when executing `soopervisor export` and putting a "resolved" `pipeline.yaml` in the Docker container. Ploomber already has a feature that allows defining placeholders in a separate file (`env.yaml`), we can use this feature to facilitate the implementation.

This  is a high level description of what I think it's the best way to do this:

[`docker.build`](https://github.com/ploomber/soopervisor/blob/d3e533761f5c26f796175254fbdd5c716e268130/src/soopervisor/commons/docker.py#L157) is the function that builds the docker image, and it already receives the path to the `pipeline.yaml` in the `entry_point` argument. We can use this value and pass it to this [function](https://github.com/ploomber/ploomber/blob/4417994432aeb92891c75111699c520f7da4c737/src/ploomber/util/default.py#L281) (note that this is defined in ploomber), which will return is what `env.yaml` file the `entry_point` will use, let's call this value `path_to_env`.

We can use this [function](https://github.com/ploomber/ploomber/blob/4417994432aeb92891c75111699c520f7da4c737/src/ploomber/env/envdict.py#L55) to get the values of the placeholders (like {{now}} and {{git_hash}}).

```python
from ploomber.env.envdict import EnvDict

env=EnvDict({})

env._data
```

Then, we apply some logic depending on the value of `path_to_env`, right before [compressing](https://github.com/ploomber/soopervisor/blob/d3e533761f5c26f796175254fbdd5c716e268130/src/soopervisor/commons/docker.py#L268) the source code.

if `path_to_env` is `None`, we create an `env.yaml` in the same directory as the `pipeline.yaml`, and write the contents of the placeholders, if `path_to_env` is not None, then we load its contents and append the values of the placeholders. However, we should not override keys (only add the ones that do not exist).

### PR DESCRIPTION
**Pull request by feribg**
## Issue ticket number and link
Closes #116

## Checklist before requesting a review
- [ x  ] I have performed a self-review of my code
- [ x ] I have added thorough tests (when necessary).
- [  ] I have added the right documentation (when needed). Product update? If yes, write one line about this update.

### CODE REVIEW

File: tests/test_commons_docker.py [hunk 1]
```diff
@@ -180,7 +180,8 @@ def test_docker_build_caches_pkg_installation(EXPORTER, config,
                          capture_output=True)
 
     ls = out.stdout.decode()
-    expected = ('environment.yml\nfast-pipeline.tar.gz\n'
+
+    expected = ('env.yaml\nenvironment.yml\nfast-pipeline.tar.gz\n'
```
Comment (edublancas) [inline tests/test_commons_docker.py:184]: can you add two more tests?

let's check the content of the env.yaml, one when a user already has one (then the content should be the existing user values + the default ones) and another one when there is no user env.yaml  and we create it

**Review by edublancas — CHANGES_REQUESTED** [review]

File: tests/test_commons_docker.py (hunk 1 above)
Comment (feribg) [reply tests/test_commons_docker.py:184]: Yes but can you please share some thoughts on how these test projects are created, it doesnt seem any of them uses a git repo etc ?

File: tests/test_commons_docker.py [hunk 2]
```diff
@@ -180,7 +182,16 @@ def test_docker_build_caches_pkg_installation(EXPORTER, config,
...
                 'fast_pipeline.py\nmy-env\npipeline.yaml\n'
                 'requirements.lock.txt\nsoopervisor.yaml\n')
+
     assert ls == expected
+
+    out = subprocess.run(['docker', 'run', 'fast-pipeline', 'cat', 'env.yaml'],
+                         check=True,
+                         capture_output=True)
+
+    env_contents = out.stdout.decode()
```
Comment (feribg) [inline tests/test_commons_docker.py:196]: @edublancas this file only has `{}` as contents instead it should have `build_time` and the git placeholders it does have the build time locally pre sending to docker and not the placeholders but im kinda lost in all the indirection where is that stuff being injected / cleared.

**Review by edublancas — APPROVED** [review]

**Comment by edublancas** [comment]
hey @feribg, I fixed and merged the PR. Please take the version in master for a spin and let me know if it works for you. If so, I'll make a release.

Note: I saw you had a `build_time` placeholder (apart from the git ones) in your logic, I deleted it because I believe that will never get executed as it is not defined as a default placeholder in ploomber. is that right?

**Comment by feribg** [comment]
@edublancas thanks i will take a look and try to use the master one and report back. Yes I just added that one but didnt end up implementing it the idea being is to act as a git hash if it's a git repo but I dont really need it, figured it might be useful.

**Comment by edublancas** [comment]
great, I also released a new ploomber version that contains [your changes](https://github.com/ploomber/ploomber/blob/master/CHANGELOG.md#0218-2022-12-27).

**Comment by feribg** [comment]
@edublancas  doesn't look like its working correctly. 

[`env_default = EnvDict({}, path_to_here=Path(path_to_env).parent)`](https://github.com/ploomber/soopervisor/blob/69963e83a8515ba1f6d6c5b9c3a1db2ec530195c/src/soopervisor/commons/docker.py#L98)

```* {{git_hash}}: Ensure git is installed and git repository exists```

```>>> import soopervisor
>>> soopervisor.__version__
'0.9.1'```

Seems like it keeps looking for .git for whatever reason and its a bit tricky to debug it in AWS batch.

**Comment by edublancas** [comment]
Since ploomber implements the logic for expanding these values, it's possible to replicate this exclusively with ploomber, right? No AWS Batch, no docker.

Based on the error message, looks like there is a bug that's erroring because there is no repository, so probably we need to add some extra logic to ploomber that checks if those placeholders are defined already and skip checking for a git repository. Do you have time to submit a PR?

**Pull request merged**